src/components/MlTemporalController/MlTemporalController.tsx

```diff
@@ -0,0 +1,355 @@
+		tempFeatures.sort((a, b) => (a.properties?.[timeField] < b.properties?.[timeField] ? -1 : 1));
+		return tempFeatures[tempFeatures.length - 1]?.properties?.[timeField] || 0;
+	}
+	return 0;
+}
+
+const MlTemporalController = (props: MlTemporalControllerProps) => {
+	const mapHook = useMap({
+		mapId: props.mapId,
+		waitForLayer: props.insertBeforeLayer,
+	});
+	const initializedRef = useRef(false);
+
+	const labelField = props.labelField || props.geojson?.features[0]?.properties?.[0] || '';
+	const minVal = React.useMemo(() => {
+		if (props.minVal) {
+			return props.minVal;
+		}
+		return getMinVal(props.geojson, props.timeField);
+	}, [props.minVal, props.geojson, props.timeField]);
+
+	const maxVal = React.useMemo(() => {
+		if (props.maxVal) {
+			return props.maxVal;
+		}
+		return getMaxVal(props.geojson, props.timeField);
+	}, [props.maxVal, props.geojson, props.timeField]);
+
+	const [type, setType] = useState(props.type || 'circle');
+	const [step, setStep] = useState(props.step || 1);
+	const [fadeIn, setFadeIn] = useState(props.fadeIn || 5);
+	const [fadeOut, setFedeOut] = useState(props.fadeOut || 5);
+
+	const [featuresColor, setFeatureColor] = useState(props.featuresColor || '#1976D2');
+	const [labels, setLabels] = useState(props.label || true);
+	const [labelColor, setlabelColor] = useState(props.labelColor || '#000');
+	const [labelFadeIn, setLabelFadein] = useState(props.labelFadeIn || 5);
+	const [labelFadeOut, setLabelFadeOut] = useState(props.labelFadeOut || 5);
+
+	const [currentVal, setCurrentVal] = useState<number>(props.initialVal || minVal);
+	const [isPlaying, setIsPlaying] = useState(false);
+	const [accumulate, setAccumulate] = useState(props.accumulate || false);
+
+	const intervalRef: any = useRef();
+
+	const paintPicker = {
+		type: type,
+		timeField: props.timeField,
+		currentVal: currentVal,
+		minVal: minVal,
+		isPlaying: isPlaying,
+		fadeIn: fadeIn,
+		fadeOut: fadeOut,
+		step: step,
+		featuresColor: featuresColor,
+		accumulate: accumulate,
+		userPaint: props.paint,
+	};
+
+	const paint = usePaintPicker(paintPicker);
+
+	useEffect(() => {
+		if (!mapHook.map || initializedRef.current) return;
+		initializedRef.current = true;
+	}, [mapHook.map, props.mapId]);
+
+	useEffect(() => {
+		return () => {
+			if (intervalRef.current) {
+				clearInterval(intervalRef.current);
+			}
+		};
+	}, []);
+
+	//use callback function from props, if exists
+	useEffect(() => {
+		if (typeof props.onStateChange === 'function') {
+			props.onStateChange(currentVal);
+		}
+	}, [props.onStateChange, currentVal]);
+
+	//get min and max time value
+	useEffect(() => {
+		if (!props.initialVal) {
+			const neuMin = getMinVal(props.geojson, props.timeField);
+			setCurrentVal(neuMin > minVal ? neuMin : minVal);
+		} else {
+			setCurrentVal(props.initialVal);
+		}
+	}, [minVal, props.geojson, props.timeField, props.initialVal]);
+
+	// filter geojson
+	const filteredData = useMemo<FeatureCollection | undefined>(() => {
```

Comment:
We should separate some of the logic into a useTemporalController hook that can be used like:

```
const {filteredData, currentVal, setCurrentVal, isPlaying, setIsPlaying, minVal, maxVal} = useTemporalController({
  geojson: props.geojson,
  timeField: props.timeField,
  minVal: props.minVal,
  maxVal: props.maxVal
});
```